src/Elasticsearch.Extensions.Logging/ElasticsearchDataShipper.cs

```diff
@@ -0,0 +1,135 @@
+			new BlockingCollection<LogEvent>(MaxQueuedMessages);
+
+		private readonly Thread _outputThread;
+
+		public ElasticsearchDataShipper()
+		{
+			_outputThread = new Thread(ProcessLogQueue) { IsBackground = true, Name = $"{nameof(ElasticsearchLogger)}.{nameof(ProcessLogQueue)}" };
+			_outputThread.Start();
+		}
+
+		private IElasticLowLevelClient _lowLevelClient = default!;
+
+		private ElasticsearchLoggerOptions _options = default!;
+
+		internal ElasticsearchLoggerOptions Options
+		{
+			get => _options;
+			set
+			{
+				_options = value;
+				UpdateClient();
+			}
+		}
+
+		public void Dispose()
+		{
+			_messageQueue?.CompleteAdding();
+			try
+			{
+				_outputThread.Join(1500); // with timeout in case writer is locked
+			}
+			catch (ThreadStateException) { }
+
+			_messageQueue?.Dispose();
+		}
+
+		public void EnqueueMessage(LogEvent logEvent)
+		{
+			if (!_messageQueue.IsAddingCompleted)
+			{
+				try
+				{
+					_messageQueue.Add(logEvent);
+					return;
+				}
+				catch (InvalidOperationException) { }
+			}
+
+			// Adding is complete, so just log the message
+			try
+			{
+				PostEvent(logEvent);
+			}
+			catch (Exception) { }
+		}
+
+		private void PostEvent(LogEvent logEvent)
+		{
+			var indexTime = logEvent.Timestamp ?? ElasticsearchLoggerProvider.LocalDateTimeProvider();
+			if (_options.IndexOffset.HasValue) indexTime = indexTime.ToOffset(_options.IndexOffset.Value);
+
+			var index = string.Format(_options.Index, indexTime);
+
+			var localClient = _lowLevelClient;
+			var response = localClient.Index<StringResponse>(index,
+				PostData.StreamHandler(logEvent,
+					(@event, stream) => logEvent.Serialize(stream),
+					// async variant not used yet but will when we move to channels/tpl in the future
+					async (@event, stream, ctx) => await logEvent.SerializeAsync(stream, ctx).ConfigureAwait(false)
+					)
+				);
+		}
+
+		private void ProcessLogQueue()
+		{
+			try
+			{
+				foreach (var logEvent in _messageQueue.GetConsumingEnumerable()) PostEvent(logEvent);
+			}
+			catch
+			{
+				try
+				{
+					_messageQueue.CompleteAdding();
+				}
+				catch { }
+			}
+		}
+
+		private void UpdateClient()
+		{
+			// TODO: Check if Uri has changed before recreating
+			// TODO: Injectable factory? Or some way of testing.
+			var connectionPool = _options.ShipTo.CreateConnectionPool();
+			var nodes = _options.ShipTo.NodeUris.ToArray();
+
+			ConnectionConfiguration settings;
+
+			if (nodes.Length == 0 && _options.ShipTo.ConnectionPoolType != ConnectionPoolType.Cloud)
+			{
+				// This is SingleNode with "http://localhost:9200"
+				settings = new ConnectionConfiguration();
+			}
+			else if (_options.ConnectionPoolType == ConnectionPoolType.SingleNode
+				|| _options.ConnectionPoolType == ConnectionPoolType.Unknown && nodes.Length == 1)
+				settings = new ConnectionConfiguration(nodes[0]);
+			else
+			{
+				settings = new ConnectionConfiguration(connectionPool);
```

Comment:
Looks like `_options.ShipTo.CreateConnectionPool();` can return `null`: 

https://github.com/elastic/ecs-dotnet/blob/249b86d9163bd373182ef5d458910a98f7a7973e/src/Elasticsearch.Extensions.Logging/ShipTo.cs#L61-L86

but `ConnectionConfiguration` does not guard against `null`: 

https://github.com/elastic/elasticsearch-net/blob/f1ffbf260132eb83ea4e7950c1bcf19ef679ee6f/src/Elasticsearch.Net/Configuration/ConnectionConfiguration.cs#L210

Reply:
Good catch!, this is a carry over from #98, will update the meta ticket #95  to follow up in a separate PR